locks: introduce mutex for tx download, flush rejection filters once per tip change

glozow opened the pull request:
See #27463 for full project tracking.

This contains the first few commits of #30110, which require some thinking about thread safety in review.
- Introduce a new `m_tx_download_mutex` which guards the transaction download data structures including `m_txrequest`, the rolling bloom filters, and `m_orphanage`. Later this should become the mutex guarding `TxDownloadManager`.
  - `m_txrequest` doesn't need to be guarded using `cs_main` anymore
  - `m_recent_confirmed_transactions` doesn't need its own lock anymore
  - `m_orphanage` doesn't need its own lock anymore
- Adds a new `ValidationInterface` event, `ActiveTipChanged`, which is a synchronous callback whenever the tip of the active chainstate changes.
- Flush `m_recent_rejects` and `m_recent_rejects_reconsiderable` on `ActiveTipChanged` just once instead of checking the tip every time `AlreadyHaveTx` is called. This should speed up calls to that function (no longer comparing a block hash each time) and removes the need to lock `cs_main` every time it is called.

Motivation:
- These data structures need synchronization. While we are holding `m_tx_download_mutex`, these should hold:
  - a tx hash in `m_txrequest` is not also in `m_orphanage`
  - a tx hash in `m_txrequest` is not also in `m_recent_rejects` or `m_recent_confirmed_transactions`
  - In the future, orphan resolution tracking should also be synchronized. If a tx has an entry in the orphan resolution tracker, it is also in `m_orphanage`, and not in `m_txrequest`, etc.
- Currently, `cs_main` is used to e.g. sync accesses to `m_txrequest`. We should not broaden the scope of things it locks.
- Currently, we need to know the current chainstate every time we call `AlreadyHaveTx` so we can decide whether we should update it. Every call compares the current tip hash with `hashRecentRejectsChainTip`. It is more efficient to have a validation interface callback that updates the rejection filters whenever the chain tip changes.

glozow committed:
guard TxRequest and rejection caches with new mutex

We need to synchronize between various tx download structures.
TxRequest does not inherently need cs_main for synchronization, and it's
not appropriate to lock all of the tx download logic under cs_main.

glozow committed:
add ValidationInterface::ActiveTipChange

This is a synchronous callback notifying clients of all tip changes.

It allows clients to respond to a new block immediately after it is
connected. The synchronicity is important for things like
m_recent_rejects, in which a transaction's validity can change (rejected
vs accepted) when this event is processed. For example, the transaction
might have a timelock condition that has just been met. This is distinct
from something like m_recent_confirmed_transactions, in which the
validation outcome is the same (valid vs already-have), so it does not
need to be reset immediately.

glozow committed:
remove obsoleted hashRecentRejectsChainTip

This also means AlreadyHaveTx no longer needs cs_main held.

glozow committed:
lock m_recent_confirmed_transactions using m_tx_download_mutex

hebasto commented on src/net_processing.cpp:
```diff
@@ -3181,9 +3181,10 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
     AssertLockNotHeld(m_peer_mutex);
     AssertLockHeld(g_msgproc_mutex);
-    AssertLockHeld(cs_main);
+    /* AssertLockHeld(cs_main); */
```
This commented line maybe requires more explanation?

glozow replied:
I missed deleting that. Fixed, thanks!

hebasto commented on src/net_processing.cpp:
```diff
@@ -2281,26 +2294,13 @@ void PeerManagerImpl::BlockChecked(const CBlock& block, const BlockValidationSta
     if (m_orphanage.HaveTx(gtxid)) return true;
 
     if (include_reconsiderable && m_recent_rejects_reconsiderable.contains(hash)) return true;
 
-    {
-        LOCK(m_recent_confirmed_transactions_mutex);
-        if (m_recent_confirmed_transactions.contains(hash)) return true;
-    }
+    if (m_recent_confirmed_transactions.contains(hash)) return true;
```
It seems reasonable to prepend this function body with `AssertLockHeld(m_tx_download_mutex);` now, no?

glozow replied:
added

hebasto commented on src/validation.cpp:
```diff
@@ -3407,7 +3408,11 @@ bool Chainstate::ActivateBestChain(BlockValidationState& state, std::shared_ptr<
-        }
+            } // release MempoolMutex
+            if (m_chainman.m_options.signals) {
+                m_chainman.m_options.signals->UpdatedBlockTipSync(pindexNewTip);
```
This call invokes logging, i.e. I/O operations, while holding the `::cs_main` and `m_mempool->cs` mutexes. It is considered as a suboptimal approach from the performance point of view. Could it be avoided?

glozow replied:
I don't think it's true that `m_mempool->cs` is still held here (see line above)?

glozow replied:
As for `cs_main`, it seems necessary for the chain tip to be what we think it is when the callback happens. IIUC the logging only happens in debug?

hebasto commented on src/net_processing.cpp:
```diff
@@ -779,7 +781,10 @@ class PeerManagerImpl final : public PeerManager
-    TxRequestTracker m_txrequest GUARDED_BY(::cs_main);
+
+    /** Protects tx download including TxRequest, rejection filter. */
+    mutable Mutex m_tx_download_mutex;
```
Is the `mutable` keyword really required?

glozow replied:
Removed now, thanks

ajtowns commented on src/net_processing.cpp:
```diff
@@ -3197,9 +3199,9 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
 void PeerManagerImpl::ProcessInvalidTx(NodeId nodeid, const CTransactionRef& ptx, const TxValidationState& state,
                                        bool maybe_add_extra_compact_tx)
 {
+    AssertLockHeld(m_tx_download_mutex);
```
Do `AssertLockNotHeld` before the `AssertLockHeld` calls?

ajtowns commented on src/validationinterface.cpp:
```diff
@@ -183,6 +183,12 @@ void ValidationSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlo
                           fInitialDownload);
 }
 
+void ValidationSignals::UpdatedBlockTipSync(const CBlockIndex *pindexNew)
```
Doesn't this also need to be called from `InvalidateBlock()`, potentially each time the tip is updated?

ajtowns reviewed (commented):
I'm not sure how to be confident that the `UpdateBlockTipSync` / `hashRecentRejectsChainTip` would be correct and didn't miss an edge case. Rest looks good to me.

glozow commented on src/net_processing.cpp:
```diff
@@ -1075,7 +1074,7 @@ class PeerManagerImpl final : public PeerManager
 
     /** Storage for orphan information */
-    TxOrphanage m_orphanage;
+    TxOrphanage m_orphanage GUARDED_BY(m_tx_download_mutex);
```
Alternatively I can add a commit that removes the `TxOrphanage` lock?

glozow replied:
Removed and added a scripted-diff commit to rename `EraseTxNoLock` to `EraseTxInternal`

glozow committed:
remove obsoleted TxOrphanage::m_mutex

The TxOrphanage is now guarded externally by m_tx_download_mutex.

glozow committed:
update recent_rejects filters on ActiveTipChange

Resetting m_recent_rejects once per block is more efficient than
comparing hashRecentRejectsChainTip with the chain tip every time we
call AlreadyHaveTx. We keep hashRecentRejectsChainTip for now to assert
that updates happen correctly; it is removed in the next commit.

glozow commented:
> I'm not sure how to be confident that the UpdateBlockTipSync / hashRecentRejectsChainTip would be correct and didn't miss an edge case. 

I (locally) split the commit into (1) update on validation interface callback and asserting that `hashRecentRejectsChainTip` is equal to the chain tip whenever we call `AlreadyHaveTx` + (2) removing `hashRecentRejectsChainTip`. I think if we see that everything runs fine with (1) it's probably correct.

I've (locally) hit a bug though, so will figure out what's wrong and then push.

ajtowns commented:
> > I'm not sure how to be confident that the UpdateBlockTipSync / hashRecentRejectsChainTip would be correct and didn't miss an edge case.
> 
> I (locally) split the commit into (1) update on validation interface callback and asserting that `hashRecentRejectsChainTip` is equal to the chain tip whenever we call `AlreadyHaveTx` + (2) removing `hashRecentRejectsChainTip`. I think if we see that everything runs fine with (1) it's probably correct.
> 
> I've (locally) hit a bug though, so will figure out what's wrong and then push.

FWIW I tried something similar, and got an assertion failure in one of the mempool functional tests (maybe mempool_reorg, and thus due to an invalidateblock call?), but it wasn't reliable, it only occurred when I ran many tests in parallel, not when I reran that test on its own. The bug will only trigger if the tip is updated via some other path and AlreadyHave is called before the tip is further updated via the expected path, so (1) doesn't seem like a terribly reliable check to me.

glozow commented:
Was it mempool_packages.py maybe? Mine tripped there on `invalidateblock` when I was adding `UpdatedBlockTip` to `InvalidateBlock`, and it was a lock ordering problem.

Hm. The alternative is to hold `cs_main` and tell txdownloadman the current chain tip pretty much all the time...

ajtowns commented:
> > FWIW I tried something similar, and got an assertion failure in one of the mempool functional tests (maybe mempool_reorg, and thus due to an invalidateblock call?)
> 
> Was it mempool_packages.py maybe? Mine tripped there on `invalidateblock` when I was adding `UpdatedBlockTip` to `InvalidateBlock`, and it was a lock ordering problem. My hierarchy is cs_main -> tx_download_mutex -> mempool.cs.

Ah, yes, I think it was.

instagibbs commented:
thanks for splitting this up, next on my review docket

instagibbs reviewed (commented):
looking pretty straightforward but I'm not an expert in the current locking setup

will give another pass in a bit

glozow commented on src/net_processing.cpp:
```diff
@@ -493,9 +493,9 @@ class PeerManagerImpl final : public PeerManager
     /** Overridden from CValidationInterface. */
     void BlockConnected(ChainstateRole role, const std::shared_ptr<const CBlock>& pblock, const CBlockIndex* pindexConnected) override
-        EXCLUSIVE_LOCKS_REQUIRED(!m_recent_confirmed_transactions_mutex);
+        EXCLUSIVE_LOCKS_REQUIRED(!m_recent_confirmed_transactions_mutex, !m_tx_download_mutex);
```
Oh, I didn't know you could do both. Happy to change if people want, but generally prefer to follow convention of the existing annotations.

ajtowns replied:
`EXCLUSIVE_LOCKS_REQUIRED(!foo)` is a stronger assertion; it says that if the caller can see the lock, it also has to have the same assertion. `LOCKS_EXCLUDED(foo)` just says you can't `LOCK()` or have `EXCLUSIVE_LOCKS_REQUIRED(foo)` -- so there's nothing prevent the caller's caller from having taken the lock.

glozow replied:
Ah that makes sense. I will leave this as is.

instagibbs replied:
ok, makes sense why I only see it used in `cs_main` contexts, since it's a global mutex. thanks

glozow committed:
[refactor] delete EraseTxNoLock, just use EraseTx

glozow commented on src/txorphanage.h:
```diff
@@ -99,7 +99,7 @@ class TxOrphanage {
 
     /** Erase an orphan by wtxid */
-    int EraseTxNoLock(const Wtxid& wtxid);
+    int EraseTxInternal(const Wtxid& wtxid);
```
Oh duh! Swapped out the last commit for just a deletion of `EraseTxNoLock`, now having all functions call `EraseTx` instead.

glozow commented on src/test/orphanage_tests.cpp:
```diff
@@ -21,15 +21,13 @@ BOOST_FIXTURE_TEST_SUITE(orphanage_tests, TestingSetup)
 class TxOrphanageTest : public TxOrphanage
 {
 public:
-    inline size_t CountOrphans() const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
```
done

instagibbs commented:
suggested changes were done, reviewed via `git range-diff master 7c3fb97 ef8de26`

instagibbs reviewed (approved):
utACK ef8de26be5478729328ac9d8a9ad6898351552b6

glozow commented on src/net_processing.cpp:
```diff
@@ -779,7 +779,10 @@ class PeerManagerImpl final : public PeerManager
     CTxMemPool& m_mempool;
-    TxRequestTracker m_txrequest GUARDED_BY(::cs_main);
+
+    /** Protects tx download including TxRequestTracker, rejection filters, and TxOrphanage. */
```
fixed, thanks

glozow reviewed (commented):
rebased

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/30111/commits/a9a6de5a7c8c411519622a32bd4b998df5d7d883

Only relevant change is suggested commit message fix

reviewed via `git range-diff master ef8de26 a9a6de5`

glozow commented:
Rebased

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/30111/commits/0e0c422aedd4009ab34eca127e4904d15e81f5be

just a rebase

reviewed via `git range-diff master a9a6de5 0e0c422`

vasild commented on src/net_processing.cpp:
```diff
@@ -2081,6 +2086,17 @@ void PeerManagerImpl::StartScheduledTasks(CScheduler& scheduler)
+{
+    AssertLockNotHeld(m_mempool.cs);
+    LOCK(m_tx_download_mutex);
+    // If the chain tip has changed, previously rejected transactions might now be invalid, e.g. due
+    // to a timelock. Reset the rejection filters to give those transactions another chance if we
+    // see them again.
```
Should `s/invalid/valid`?

vasild reviewed (approved):
ACK 0e0c422aedd4009ab34eca127e4904d15e81f5be

Changes like this are inherently difficult to review:

One has to check all variables protected by `cs_main` and ensure that none of them are required to be in sync with the ones that are removed from `cs_main` protection like `m_txrequest`.

Also, one has to check that all mutexes locked before the new one (and still locked when the new one is acquired) are never locked while the new one is locked anywhere else in the code and that mutexes locked while the new one is locked are never locked before it anywhere else.

glozow commented:
Thanks @vasild. I will take your suggestions in a followup or if I retouch this.

achow101 commented:
ACK 0e0c422aedd4009ab34eca127e4904d15e81f5be

Not as familiar with locking in their area, so not super confident on the `cs_main` related changes, but the rest (removing other mutexes and consolidating under `m_tx_download_mutex`) looks fine.

dergoegge commented on src/net_processing.cpp:
```diff
@@ -2081,6 +2086,17 @@ void PeerManagerImpl::StartScheduledTasks(CScheduler& scheduler)
+void PeerManagerImpl::UpdatedBlockTipSync(const CBlockIndex* pindexNew)
+{
+    AssertLockNotHeld(m_mempool.cs);
+    LOCK(m_tx_download_mutex);
```
Is clearing the filters is necessary in IBD?

glozow replied:
gating on whether ibd now

dergoegge reviewed (commented):
> This is a synchronous version of UpdatedBlockTip.

Given that this is the goal, it's a little weird that `UpdatedBlockTipSync` has a different signature and is called in different locations.


If it wasn't for the zmq notifier we could probably just make `UpdateBlockTip` synchronous instead of having two versions.

glozow commented on src/net_processing.cpp:
```diff
@@ -1674,8 +1674,9 @@ void PeerManagerImpl::InitializeNode(CNode& node, ServiceFlags our_services)
         m_node_states.emplace_hint(m_node_states.end(), std::piecewise_construct, std::forward_as_tuple(nodeid), std::forward_as_tuple(node.IsInboundConn()));
+        LOCK(m_tx_download_mutex);
         assert(m_txrequest.Count(nodeid) == 0);
     }
```
added

glozow commented:
Addressed @dergoegge and @vasild comments. Ready for review again, reACKs would be appreciated!

dergoegge reviewed (approved):
Code review ACK 17d41e7a547f16f0dd3802dac73a63772ca000b2

glozow commented on src/validationinterface.h:
```diff
@@ -61,6 +61,10 @@ class CValidationInterface {
+    /**
+     * Notifies listeners any time the block chain tip changes, synchronously.
+     */
+    virtual void ActiveTipChange(const CBlockIndex* pindexNew, bool is_ibd) {};
```
will do if I retouch

glozow replied:
done

instagibbs reviewed (approved):
reACK 17d41e7a547f16f0dd3802dac73a63772ca000b2 (noting my weak ability to review locking)

biggest change is changing the new function to `ActiveTipChange` and making ibd a no-op

reviewed via `git range-diff master 0e0c422 17d41e7`

glozow commented on src/validation.cpp:
```diff
@@ -3684,6 +3690,12 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
+
+        // Fire ActiveTipChange now for the current chain tip to make sure clients are notified.
+        // ActivateBestChain may call this as well, but not necessarily.
+        if (m_chainman.m_options.signals) {
```
That seems incorrect. `ActivateBestchain` exits before `ActiveTipChange` if it realizes there is nothing to do, e.g. if `invalidateblock` only causes the tip to change from block n to block n-1.

https://github.com/bitcoin/bitcoin/blob/9b480f7a25a737c9c4ebc33401e94d66c2da9ec3/src/validation.cpp#L3505-L3508

glozow replied:
What do you mean by "here"?

Eunovo replied:
By "here", I mean in https://github.com/bitcoin/bitcoin/blob/9b480f7a25a737c9c4ebc33401e94d66c2da9ec3/src/validation.cpp#L3505-L3508

glozow replied:
Because then it would fire even when it's not supposed to fire. We're aiming to add a notification for when the tip changes.

mzumsande commented on src/net_processing.cpp:
```diff
@@ -2092,6 +2094,21 @@ void PeerManagerImpl::StartScheduledTasks(CScheduler& scheduler)
 
+void PeerManagerImpl::ActiveTipChange(const CBlockIndex* pindexNew, bool is_ibd)
+{
+    AssertLockNotHeld(m_mempool.cs);
```
I think you meant `m_tx_download_mutex`

glozow replied:
No, I think the `LOCK(m_tx_download_mutex)` already checks that it isn't already held, since it's a `Mutex`.

I guess this seems a bit random, but I'm trying to establish the lock hierarchy as `m_tx_download_mutex`, then `m_mempool.cs` (which is a `RecursiveMutex`). That's also why we can only fire `ActiveTipChanged` after releasing the mempool mutex. Later code added to `TxDownloadManager` will call `CTxMemPool` functions that take the mempool lock (see e.g. https://github.com/bitcoin/bitcoin/pull/30110/files#diff-5d7e42fb1e904f54956c0975a9307c58d8d36d77ad9da8bcd8989437356e3229R264 or grep `exists` within txdownload_impl).

glozow replied:
Ah I was wrong about `LOCK` doing the assert, added that as well.

hebasto replied:
As far as I understand, `AssertLockNotHeld(m_mempool.cs);` is an attempt to force the mutex hierarchy mentioned in here: https://github.com/bitcoin/bitcoin/blob/c85accecafc20f6a6ae94bdf6cdd3ba9747218fd/src/net_processing.cpp#L784

As `m_mempool.cs` is not used in this function, I suggest to add a comment to prevent tempting of removing this line on the "clean up" purpose in the future.

mzumsande reviewed (commented):
Concept ACK - don't know the txrequest logic very well though.

The PR description should be updated (`UpdatedBlockTipSync` was renamed).

glozow commented:
- Updated the PR title and description, thanks @mzumsande 
- Rebased for #30394
- Renamed variable https://github.com/bitcoin/bitcoin/pull/30111#discussion_r1665560023
- Added some more documentation to `m_tx_download_mutex` to address https://github.com/bitcoin/bitcoin/pull/30111#discussion_r1678336506.

instagibbs commented:
reACK c85accecafc20f6a6ae94bdf6cdd3ba9747218fd

reviewed via `git range-diff master 17d41e7 c85acce`

dergoegge reviewed (approved):
Code review ACK c85accecafc20f6a6ae94bdf6cdd3ba9747218fd

hebasto commented on src/net_processing.cpp:
```diff
@@ -501,13 +503,13 @@ class PeerManagerImpl final : public PeerManager
     bool HasAllDesirableServiceFlags(ServiceFlags services) const override;
     bool ProcessMessages(CNode* pfrom, std::atomic<bool>& interrupt) override
-        EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex, !m_recent_confirmed_transactions_mutex, !m_most_recent_block_mutex, !m_headers_presync_mutex, g_msgproc_mutex);
+        EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex, !m_most_recent_block_mutex, !m_headers_presync_mutex, g_msgproc_mutex, !m_tx_download_mutex);
```
The `PeerManagerImpl::ProcessMessages` implementation  has multiple early `return` statements. In such a case, the Developer Notes [suggest](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#threads-and-synchronization) to:
> Combine annotations in function declarations with run-time asserts in function definitions

i.e., `AssertLockNotHeld(m_tx_download_mutex);`

glozow replied:
thanks, added in #30507

hebasto commented on src/net_processing.cpp:
```diff
@@ -6281,6 +6297,7 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
         //
         // Message: getdata (transactions)
         //
+        LOCK(m_tx_download_mutex);
```
Do we need to hold `m_tx_download_mutex` for this call: https://github.com/bitcoin/bitcoin/blob/c85accecafc20f6a6ae94bdf6cdd3ba9747218fd/src/net_processing.cpp#L6327-L6328 a few lines down?

glozow replied:
no we don't, changed in #30507

theStack reviewed (approved):
Light code-review ACK c85accecafc20f6a6ae94bdf6cdd3ba9747218fd

Fwiw, I've also spun up a fresh mainnet node instance on a publicly reachable machine running this PR for the last few days (out of IBD since Friday morning) and didn't notice any problems so far.

hebasto commented:
>    * Introduce a new `m_tx_download_mutex` which guards the transaction download data structures including `m_txrequest`, the rolling bloom filters, and `m_orphanage`.
>       
>      * `m_orphanage` doesn't need its own lock anymore

In the master branch, the design of the `TxOrphanage` class is thread-safe for all purposes. I don't think that the fact that `m_orphanage` does not require additional synchronization justifies the change of `TxOrphanage` class design. The new `m_tx_download_mutex` guarantees no contention when locking `TxOrphanage::m_mutex`, which makes the latter cheap.

If still insisting on the `TxOrphanage` class change, then perhaps document that it is no longer thread-safe and requires an external synchronization?

glozow commented:
> In the master branch, the design of the TxOrphanage class is thread-safe for all purposes. I don't think that the fact that m_orphanage does not require additional synchronization justifies the change of TxOrphanage class design. 

See https://github.com/bitcoin/bitcoin/pull/30111#discussion_r1601902513 for background. `TxOrphanage` is not a multi-purpose container. It has 1 specific purpose, and its user will need to externally synchronize it with other data structures to fulfill that purpose. It seems it would just slow things down to acquire and release a lock whenever we use it.

> perhaps document that it is no longer thread-safe and requires an external synchronization?

I've opened #30507 to add documentation to txorphanage.h, as well as the other suggestions from @hebasto and @theStack (thanks!).

glozow commented on src/net_processing.cpp:
```diff
@@ -2119,20 +2118,15 @@ void PeerManagerImpl::BlockConnected(
-        }
+    for (const auto& ptx : pblock->vtx) {
+        m_txrequest.ForgetTxHash(ptx->GetHash());
+        m_txrequest.ForgetTxHash(ptx->GetWitnessHash());
```
good idea, done in #30507

hebasto reviewed (approved):
ACK c85accecafc20f6a6ae94bdf6cdd3ba9747218fd, I have reviewed the code and it looks OK.
